cirq/contrib/two_qubit_gates/gate_compilation.py

```diff
@@ -0,0 +1,379 @@
+                                                     kak_vector_to_unitary)
+
+_SingleQubitGatePair = Tuple[np.ndarray, np.ndarray]
+
+
+class TwoQubitGateCompilation(NamedTuple):
+    """Represents a compilation of a target 2-qubit with respect to a base gate.
+
+    This object encodes the relationship between 4x4 unitary operators
+
+    U_target ~ k_N · U_base · k_{N-1} · ... · k_1 · U_base · k_0
+
+    where U_target, U_base are 2-local and k_j are 1-local.
+
+    Attributes:
+        base_gate: 4x4 unitary denoting U_base above.
+        target_gate: 4x4 unitary denoting U_target above.
+        local_unitaries: Sequence of 2-tuples (k_{00},k_{01}),(k_{10},k_{11})...
+            where k_j = k_{j0} ⊗ k_{j1} in the product above. Each k_{j0},
+            k_{j1} is a 2x2 unitary.
+        actual_gate: 4x4 unitary denoting the right hand side above, ideally
+            equal to U_target.
+        success: Whether actual_gate is expected to be close to U_target.
+    """
+    base_gate: np.ndarray
+    target_gate: np.ndarray
+    local_unitaries: Tuple[_SingleQubitGatePair, ...]
+    actual_gate: np.ndarray
+    success: bool
+
+
+@attr.s(auto_attribs=True)
+class GateTabulation:
+    """A 2-qubit gate compiler based on precomputing/tabulating gate products.
+
+    """
+    base_gate: np.ndarray  # Base two qubit gate. (4x4 unitary)
+    # Sequence of KAK vectors, ideally "dense" in the Weyl chamber. Shape (N,3).
+    kak_vecs: np.ndarray
+    # Sequence of 1-local operations required to achieve a given KAK vector.
+    # Index j corresponds to KAK_vecs[j], and is of the form
+    # ( (u0[0],u1[0]), (u0[1],u1[1]), ...) where u0[k] is the kth single qubit
+    # unitary acting on qubit 0 (similarly for u1)
+    single_qubit_gates: Sequence[Sequence[_SingleQubitGatePair]]
+    max_expected_infidelity: float  # Defined using entanglement fidelity.
+    summary: str  # Text summarizing the results of the tabulation procedure.
+    # Any KAK vectors which are expected to be compilable (within infidelity
+    # max_expected_infidelity) using 2 or 3 base gates.
+    missed_points: Tuple[np.ndarray, ...]
+
+    def compile_two_qubit_gate(self,
+                               unitary: np.ndarray) -> TwoQubitGateCompilation:
+        r"""Compute single qubit gates required to compile a desired unitary.
+
+        Given a desired unitary U, this computes the sequence of 1-local gates
+        k_j such that the product
+
+        k_{n-1} A k_{n-2} A ... k_1 A k_0
+
+        is close to U. Here A is the base_gate of the tabulation.
+
+        Args:
+            unitary: Unitary (U above) to compile.
+
+        Returns:
+            A TwoQubitGateCompilation object encoding the required local
+            unitaries and resulting product above.
+        """
+        unitary = np.asarray(unitary)
+        kak_vec = kak_vector(unitary, check_preconditions=False)
+        infidelities = kak_vector_infidelity(kak_vec,
```

Comment:
These seems like a really inefficient way to find the closest point, though it's quite simple which is nice. How fast is it in practice for various max infidelities? Is it usable at max infidelity `0.001`? `0.0001`? The infidelity of the decomposition should be negligible compared to the physical gate error, so 10^{-4} is presumably a good target.

Reply:
In practice this should scale with the number of mesh points, which is scales as (1/d)^3. Here d is the euclidean lattice spacing, which is \sqrt(max_infidelity)/3.  So the total number of mesh points scales as max_infidelity^{-3/2}. 

In practice the gate tabulation takes significantly longer than the gate compilation, so I can't test this for very small max_infidelity in a reasonable amount of time. In fact, I am seeing that the tabulation scales as max_infidelity^{-2}, which is surprising and I should look into. 

At 0.01, the tabulation time is about 7 seconds and the total compilation time (per gate) is 0.0022 seconds (Sorry I haven't set up a profiler on this branch so I can't get exact timing for this line). If I push max_infidelity down to 0.005, the tabulation time is about 29.5 seconds and the total compilation time is 0.0023 s. Likewise, if I set max_infidelity to 0.004, the tabulation time is about 50 seconds and the time per gate is 0.0024 s. So, worst case, the line you mention takes about 1e-4 for max_infidelity=0.01, and I would guess it takes 1e-4 * (100^(3/2) = 0.1 s for max_infidelity=0.0001. 

From my previous profiling on my own version of the code, I recall that (at least for max_infidelity not too small) the line that takes up a majority of the time is the call to `kak_decomposition`.

Update: I was able to run the example with max_infidelity=0.001. The tabulation took 1900 seconds (~31 minutes) while the compilation took 0.0043 seconds per gate. So I suspect my previous estimate is actually rather pessimistic.

Reply:
The current implementation will not be very practical for max_infidelity much smaller then 0.01, unless we precompute the tabulation once and save it. That said, in the future we can make the compiler more sophisticated by allowing for local optimization using the tabulated point as an initial guess.